lib/ipaddr.ml

```diff
@@ -27,7 +25,23 @@ type scope =
 | Site
 | Organization
 | Global
-[@@deriving sexp]
+
+let sexp_of_scope, scope_of_sexp =
+  let open Sexplib.Sexp in
+  let lst = [
+    Point, Atom "Point" ;
+    Interface, Atom "Interface";
+    Link, Atom "Link" ;
+    Admin, Atom "Admin" ;
+    Site, Atom "Site" ;
+    Organization, Atom "Organization" ;
+    Global, Atom "Global"
+  ]
+  in
+  (fun scope -> List.assoc scope lst),
+  (fun sexp -> fst (List.find (fun (_, sexp') -> match sexp, sexp' with
```

Comment:
This is going to leak a `Not_found` exception whereas the old ppx converters would throw a different one I think...

Reply:
yes.  does it matter? (it would imho if the exception would be documented) -- also I wasn't able to find any users of `scope_of_sexp`.

Reply:
I think it's definitely worth not raising more random Not_found exceptions.

Thinking about this more, why don't we take a similar approach to https://github.com/mirage/ocaml-uri/pull/121 here, and just shift out the sexp converter to a separate findlib package?  I don't like these manually written converters with subtly different semantics to the ppx ones...

Reply:
Imo the separate packages are fairly useless because they aren't going to work for abstract types without silly hacks like adding private modules everywhere.

Reply:
dear @avsm, thanks for your comment.  the whole reason for this PR is - as mentioned above - that while testing arbitrary compiler (pre-)releases and pull requests, often the ocaml-migrate-parsetree - which should offer compiler-version-independent-ppx - is not available on any of these.

moving the ppx to a sublibrary does not solve this problem (as mentioned, mirage-runtime depends on ipaddr and the `to/of_sexp` functions.

to address your `not raising more random .. exceptions` point, please take a look at how `Ipaddr.V4.t_of_sexp` is implemented atm in master (similar to `V6.t_of_sexp`):
```OCaml
  let t_of_sexp i =
    match i with
    | Sexplib.Sexp.Atom i -> of_string_exn i
    | _ -> raise (Failure "Ipaddr.V4.t: Unexpected non-atom in sexp")
```

In the current state these hand-crafted functions already raise some exceptions, `Failure` or `Parse_error` (in contrast to `Sexplib0__Sexp.Of_sexp_error` which I get from derived sexp converters).

of course we could change this to use `[@@deriving sexp]` - which would be a breaking change since the marshalled data at the moment is `Atom "127.0.0.1"`, whereas a `type t = int32 [@@deriving sexp]` would marshal the same IP to `Atom "2130706433"`.

not clear to me how to move forward here.

Reply:
> In the current state these hand-crafted functions already raise some (undocumened!) exceptions, Failure or Parse_error (in contrast to Sexplib0__Sexp.Of_sexp_error which I get from derived sexp converters).

Indeed! This is much worse than using the automatic generators.The way to move forward seems to be by checking in the generated code. That can be done from a stable release of the compiler and eliminates the need for ppx at build time.  However @rgrinberg -- I can't see any promotion rules in the repository you linked to.  Does json-derivers do the checking in of the ppx outputs by hand?

But also, we can make a better effort to keep ocaml-migrate-parsetree working with ocaml-trunk. The CI now exists for trunk snapshots, so we could propose a Travis check for ocaml trunk top ensure that an omp commit accompanies an OCaml AST change.

Reply:
@avsm not sure about your plans, esp. in respect to backwards compatibility -- I'd claim there are any number of serialised IP addresses written on disks, now if you want to "checking in the generated code", this sounds like it will break all of these serialised IP addresses (which is IMHO bad and not necessary).

I don't have much interest in checking in generated code - neither do I think this is a very sensible idea (just look back at `_oasis` generated files), but should rather be done at release time (no, I don't have enough interest in this to research how to best do this with any build system).  Also, I've limited interest in "make a better effort to keep ocaml-migrate-parsetree working with ocaml-trunk".

Reply:
@rgrinberg is that workflow documented/demonstrated anywhere? it sounds interesting and it's the first I'm hearing of it (probably my own fault, since I don't watch discuss.ocaml.org or development on dune very closely)

Reply:
I don't think it's documented anywhere, but it's the same workflow that janestreet uses to use deriving in base but stay dependency free. There's some discussion to make this more widely available (extend it to all ppx converters), so that would make the ergonomics change a little bit. You can see where I'm going with this in this thread: https://github.com/ocaml/dune/issues/1002